File: SimEnvRL/UserDataDisplay/printSummary.py
from ..generalScripts import *
import numpy as np

def printSummary(env):
    param = env.param
    initialValue = env.initialValue
    adimensionalSolution = {
        "x": env.timeHistory.T,
        "y": env.fullStateHistory.T,
        "controlAction": env.controlActionHistory_L.T
    }
    OBoptimalTrajectory = env.OBoptimalTrajectory
    # Calculations
    _, _, relativeState_L = OBNavigation(initialValue.targetState_S, initialValue.chaserState_S, param)
    if relativeState_L[1] <= 0:
        rP = "BEHIND"
    else:
        rP = "IN FRONT OF"

    gg = (relativeState_L[3:6] @ relativeState_L[0:3]) / (np.linalg.norm(relativeState_L[3:6]) * np.linalg.norm(relativeState_L[0:3]))
    if gg > 0:
        rV = "AWAY FROM"
    else:
        rV = "TOWARDS"
    anglo = np.arccos(gg)

    relDynami = np.zeros((len(adimensionalSolution["x"]), 6))
    soluz = np.array(adimensionalSolution["y"]).T
    for idx in range(len(adimensionalSolution["x"])):
        rotatedRelativeState = ReferenceFrames.convert_S_to_LVLH(soluz[idx, 0:6], soluz[idx, 6:12] - soluz[idx, 0:6], param)
        relDynami[idx, 0:6] = rotatedRelativeState

    controlAction = adimensionalSolution["controlAction"]
    relDynami = relDynami * param.xc
    if param.phaseID == 1:
        positionError = np.linalg.norm(relDynami[-1, 0:3] - param.holdingState[0:3] * param.xc) * 1e3
        velocityError = np.linalg.norm(relDynami[-1, 3:6] - param.holdingState[3:6] * param.xc) * 1e3 / param.tc
    elif param.phaseID == 2:
        positionError = np.linalg.norm(relDynami[-1, 0:3] - param.dockingState[0:3] * param.xc) * 1e3
        velocityError = np.linalg.norm(relDynami[-1, 3:6] - param.dockingState[3:6] * param.xc) * 1e3 / param.tc

    # Prints
    print("#############################")

    print(f"\nPhaseID: {param.phaseID}")
    print(f"Seed imposed: ",initialValue.seedValue,"")
    print(f"Simulated Time: {param.tspan[1] / 3600 * param.tc:.2f} [hours]")
    if OBoptimalTrajectory:
        print(f"Optimal Trajectory TOF estimated: {OBoptimalTrajectory.t[-1] / 3600 * param.tc:.2f} [hours]")
    print(f"Initial Distance between C and T: {np.linalg.norm(initialValue.DeltaIC_S[0:3]) * param.xc:.2f} [km]")
    print(f"Initial Relative velocity between C and T: {np.linalg.norm(initialValue.DeltaIC_S[3:6]) * param.xc / param.tc * 1e3:.2f} [m/s]")
    print(f"Initially: Chaser is *{rP}* the Target and moving *{rV}* the Target at an angle of {np.rad2deg(anglo):.2f} [deg]")

    print("\n#############################\n")

    print(f"Maximum Thrust Available: {param.maxAdimThrust} [-]")
    print(f"Maximum Thrust Required : {np.max(controlAction):g} [-]")
    print(f"Maximum Thrust Required (norm) : {np.max(np.linalg.norm(controlAction, axis=1)):g} [-]")

    print(f"Actual Final Position Error:    {positionError:g} [m]")
    print(f"Actual Final Velocity Error:    {velocityError:g} [m/s]")

    print("\n#################################")
    print("############## END ##############")
    print("#################################")

File: matlabScripts/MATLAB/plot/printSummary.m
% calcoli
[~,~,relativeState_L] = OBNavigation(initialStateTarget_S,initialStateChaser_S,param);
if relativeState_L(2) <= 0
    rP = "BEHIND";
else
    rP = "IN FRONT OF";
end
gg = (relativeState_L(4:6)'*relativeState_L(1:3)./norm(relativeState_L(4:6))/norm(relativeState_L(1:3)));
if gg > 0
    rV = "AWAY FROM";
else
    rV = "TOWARDS";
end
anglo = acos(gg);

relDynami = zeros(length(adimensionalSolution.x),6);
soluz = adimensionalSolution.y';
for id = 1:length(adimensionalSolution.x) 
    [rotatedRelativeState] = convert_S_to_LVLH(soluz(id,1:6)',soluz(id,7:12)'-soluz(id,1:6)',param);
    relDynami(id,1:6) = rotatedRelativeState;
end

controlAction = adimensionalSolution.controlAction;
relDynami = relDynami.*param.xc;
switch phaseID
    case 1
        positionError = norm(relDynami(end,1:3)'-param.holdingState(1:3)*param.xc)*1e3;
        velocityError = norm(relDynami(end,4:6)'-param.holdingState(4:6)*param.xc)*1e3/param.tc;
    case 2
        positionError = norm(relDynami(end,1:3)'-param.dockingState(1:3)*param.xc)*1e3;
        velocityError = norm(relDynami(end,4:6)'-param.dockingState(4:6)*param.xc)*1e3/param.tc;
end

%% PRINTS
fprintf("#############################\n")

fprintf("\nPhaseID: %d\n",phaseID);
fprintf("Seed used: %s\n", num2str(param.rng_settings.Seed))
fprintf("Simulated Time: %.2f [hours]\n",adimensionalSolution.x(end)/3600*param.tc);
if ~isempty(OBoptimalTrajectory)
    fprintf(" Optimal Trajectory TOF estimated: %.2f [hours]\n",OBoptimalTrajectory.t(end)/3600*param.tc)
end
fprintf("Initial Distance between C and T: %.2f [km]\n",norm(DeltaIC_S(1:3))*param.xc);
fprintf("Initial Relative velocity between C and T: %.2f [m/s]\n",norm(DeltaIC_S(4:6))*param.xc/param.tc*1e3);
fprintf("Initially: Chaser is *%s* the Target and moving *%s* the Target at an angle of %.2f [deg]\n",rP,rV,rad2deg(anglo));

fprintf("\n#############################\n\n")

fprintf("Maximum Thrust Available: %g [-]\n",param.maxAdimThrust);
fprintf("Maximum Thrust Required : %g [-]\n",max(max(controlAction)));
fprintf("Maximum Thrust Required (norm) : %g [-]\n\n",max(norm(controlAction,1)));

fprintf("Actual Final Position Error:    %g [m]\n",positionError);
fprintf("Actual Final Velocity Error:    %g [m/s]\n",velocityError);

fprintf("\n#############################\n\n")

fprintf("TERMINAL STATE:\n")
fprintf("  position R: %f [m]\n",  relDynami(end,1)*1e3)
fprintf("  position V: %f [m]\n",  relDynami(end,2)*1e3)
fprintf("  position H: %f [m]\n",  relDynami(end,3)*1e3)
fprintf("  velocity R: %f [m/s]\n",relDynami(end,4)*1e3/param.tc)
fprintf("  velocity V: %f [m/s]\n",relDynami(end,5)*1e3/param.tc)
fprintf("  velocity H: %f [m/s]\n",relDynami(end,6)*1e3/param.tc)

fprintf("\n");
if info.aimReachedBool
    fprintf("\n <!>>>>> DOCKING SUCCESSFUL <<<<<<!> \n\n")
elseif info.crashedBool
    fprintf("\n <!!!!><><><> CRASHED <><><><!!!!> \n\n")
else
    fprintf("\n SIMULATION RUN OUT OF TIME \n")
end
fprintf("\n#################################\n")
fprintf("############## END ##############\n")
fprintf("#################################\n")